From d58f1426fb4835bc05dda4f44d53a4386ac3113a Mon Sep 17 00:00:00 2001
From: "Travis A. Everett" <travis.a.everett@gmail.com>
Date: Thu, 21 Sep 2023 21:45:20 -0500
Subject: define _PyVerify_fd

ActiveState's python, which we're using in nixpgks, removed the macro
for _PyVerify_fd from Include/fileobject.h in commit e361063c. Oil's
~forked + modified copy of posixmodule.c still includes this function.

I could in theory try to follow the rest of ActiveState's changes here
but the diff between ActiveState's copy _before_ this change and Oil's
is over 8k lines, so it'd take a good bit of work just to figure out
if the diff is entirely stuff that oil dropped because it doesn't need
it or if the code has otherwise meaningfully diverged.
---
 pyext/posixmodule.c | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/pyext/posixmodule.c b/pyext/posixmodule.c
index 7478d64bf..3c5ab9a51 100644
--- a/pyext/posixmodule.c
+++ b/pyext/posixmodule.c
@@ -1,4 +1,3 @@
-
 /* OVM_MAIN: POSIX module implementation stripped down for Oil */
 
 #define PY_SSIZE_T_CLEAN
@@ -6,6 +5,20 @@
 #include "Python.h"
 #include "structseq.h"
 
+#if defined _MSC_VER && _MSC_VER >= 1400
+ /* A routine to check if a file descriptor is valid on Windows.  Returns 0
+  * and sets errno to EBADF if it isn't.  This is to avoid Assertions
+  * from various functions in the Windows CRT beginning with
+  * Visual Studio 2005
+  */
+ int _PyVerify_fd(int fd);
+ #elif defined _MSC_VER && _MSC_VER >= 1200
+ /* fdopen doesn't set errno EBADF and crashes for large fd on debug build */
+ #define _PyVerify_fd(fd) (_get_osfhandle(fd) >= 0)
+ #else
+ #define _PyVerify_fd(A) (1) /* dummy */
+ #endif
+
 #ifdef __cplusplus
 extern "C" {
 #endif
-- 
2.39.0

